src/transform/src/predicate_pushdown.rs

```diff
@@ -581,52 +567,142 @@ impl PredicatePushdown {
-                                push_downs[input1].push(MirScalarExpr::CallBinary {
-                                    func: BinaryFunc::Or,
-                                    expr1: Box::new(MirScalarExpr::CallBinary {
-                                        func: BinaryFunc::Eq,
-                                        expr1: Box::new(expr2.clone()),
-                                        expr2: Box::new(expr1.clone()),
-                                    }),
-                                    expr2: Box::new(MirScalarExpr::CallBinary {
-                                        func: BinaryFunc::And,
-                                        expr1: Box::new(MirScalarExpr::CallUnary {
-                                            func: UnaryFunc::IsNull,
-                                            expr: Box::new(expr2.clone()),
+                        // class. Push a predicate for every pair of expressions
+                        // in the equivalence that etiher belong to a single
+                        // input or can be localized to a given input through
+                        // the rest of equivalences.
+                        let mut to_remove = Vec::new();
+                        for input in 0..inputs.len() {
+                            // Vector of pairs (position within the equivalence, localized
+                            // expression). The position is None for expressions derived through
+                            // other equivalences.
+                            let localized = equivalences[equivalence_pos]
+                                .iter()
+                                .enumerate()
+                                .filter_map(|(pos, expr)| {
+                                    if let MirScalarExpr::Column(col_pos) = &expr {
+                                        let local_col = input_mapper.map_column_to_local(*col_pos);
+                                        if input == local_col.1 {
+                                            return Some((
+                                                Some(pos),
+                                                MirScalarExpr::Column(local_col.0),
+                                            ));
+                                        } else {
+                                            return None;
+                                        }
+                                    }
+                                    let mut inputs = input_mapper.lookup_inputs(expr);
+                                    if let Some(single_input) = inputs.next() {
+                                        if input == single_input && inputs.next().is_none() {
+                                            return Some((
+                                                Some(pos),
+                                                input_mapper.map_expr_to_local(expr.clone()),
+                                            ));
+                                        }
```

Comment:
Do these lines exist to avoid unnecessarily cloning `equivalences` ?

Reply:
Yes, but also to detect whether the expression was already a pushable to the current input or was made `pushable` after localizing it. In the former case, we mark the expression in the equivalence as removable if it is pushed down below within a predicate, but in the latter we leave it there. If the expression belongs to a different input, it will be marked as `removable` in that pass.

Now that I go through it again... That is not 100% correct, since it may still leave equivalences that are fully derivable from other equivalences, like in the following case, where the last equivalence could be removed safely without introducing any cross join, and eventually, removed completely, since it adds nothing:

```
materialize=> explain select * from t1, t2 where t1.f1 = t2.f1 and t1.f2 = t2.f2 and t1.f1 + t2.f2 = t2.f1 + t1.f2;
                         Optimized Plan                         
----------------------------------------------------------------
 %0 =                                                          +
 | Get materialize.public.t1 (u756)                            +
 | ArrangeBy (#0, #1)                                          +
                                                               +
 %1 =                                                          +
 | Get materialize.public.t2 (u758)                            +
 | ArrangeBy (#0, #1)                                          +
                                                               +
 %2 =                                                          +
 | Join %0 %1 (= #0 #2) (= #1 #3) (= (#0 + #3) (#2 + #1))      +
 | | implementation = DeltaQuery                               +
 | |   delta %0 %1.(#0, #1)                                    +
 | |   delta %1 %0.(#0, #1)                                    +
 | | demand = (#0, #1)                                         +
 | Filter !(isnull(#0)), !(isnull(#1)), ((#0 + #1) = (#0 + #1))+
 | Project (#0, #1, #0, #1)                                    +
 
(1 row)

materialize=> 
```

Reply:
Actually, the comment was right, but the code wasn't. With the following patch, the redundant equivalence can be safely removed:

```
diff --git a/src/transform/src/predicate_pushdown.rs b/src/transform/src/predicate_pushdown.rs
index 0dc0cc5fd..af07fbdba 100644
--- a/src/transform/src/predicate_pushdown.rs
+++ b/src/transform/src/predicate_pushdown.rs
@@ -686,7 +686,7 @@ impl PredicatePushdown {
                                     // The equivalence is either a single input one or fully localizable
                                     // to a single input through other equivalences, so it can be removed
                                     // completely without introducing any new cross join.
-                                    to_remove.extend(localized.iter().filter_map(|(pos, _)| *pos));
+                                    to_remove.extend(0..equivalences[equivalence_pos].len());
                                 } else {
                                     // Leave an expression from this input in the equivalence to avoid
                                     // cross joins
```

```
materialize=> explain select * from t1, t2 where t1.f1 = t2.f1 and t1.f2 = t2.f2 and t1.f1 + t2.f2 = t2.f1 + t1.f2;
            Optimized Plan             
---------------------------------------
 %0 =                                 +
 | Get materialize.public.t1 (u756)   +
 | ArrangeBy (#0, #1)                 +
                                      +
 %1 =                                 +
 | Get materialize.public.t2 (u758)   +
 | ArrangeBy (#0, #1)                 +
                                      +
 %2 =                                 +
 | Join %0 %1 (= #0 #2) (= #1 #3)     +
 | | implementation = DeltaQuery      +
 | |   delta %0 %1.(#0, #1)           +
 | |   delta %1 %0.(#0, #1)           +
 | | demand = (#0, #1)                +
 | Filter !(isnull(#0)), !(isnull(#1))+
 | Project (#0, #1, #0, #1)           +
 
(1 row)

materialize=> 
```